modules/apps/transfer/types/transfer_authorization.go

```diff
@@ -145,6 +157,43 @@ func isAllowedAddress(ctx sdk.Context, receiver string, allowedAddrs []string) b
 	return false
 }
 
+// areAllowedPacketDataKeys returns a boolean indicating if the memo is valid for transfer. If it is not valid and non-nil error is also returned.
+func areAllowedPacketDataKeys(ctx sdk.Context, memo string, allowedPacketDataList []string) (bool, error) {
+	// if the allow list is empty, then the memo must be an empty string
+	if len(allowedPacketDataList) == 0 {
+		return len(strings.TrimSpace(memo)) == 0, nil
+	}
+
+	// if allowedPacketDataList has only 1 element and it equals AllowAllPacketDataKeys
+	// then accept all the packet data keys
+	if len(allowedPacketDataList) == 1 && allowedPacketDataList[0] == AllowAllPacketDataKeys {
+		return true, nil
+	}
+
+	// unmarshal memo
+	jsonObject := make(map[string]interface{})
+	err := json.Unmarshal([]byte(memo), &jsonObject)
+	if err != nil {
+		return false, err
+	}
+
+	gasCostPerIteration := ctx.KVGasConfig().IterNextCostFlat
+
+	for _, key := range allowedPacketDataList {
+		_, ok := jsonObject[key]
+		if ok {
+			ctx.GasMeter().ConsumeGas(gasCostPerIteration, "transfer authorization")
+			delete(jsonObject, key)
+		}
+	}
+
+	if len(jsonObject) != 0 {
+		return false, fmt.Errorf("not allowed packet data key")
+	}
```

Comment:
Nice way of turning around the iteration and solving the issue, @GNaD13! :)

Maybe it would be nice if you could add to the error message the keys that are left in `jsonObject`, so that it's easier to identify what keys are not valid.

Reply:
Thanks for your suggestion @crodriguezvega. I've updated the code